### ISSUE TITLE
Implement Dense-Sparse-Dense (DSD) Training, or, freezing lots of weights

### ISSUE THREAD

**Issue by cheshirecats** [apache/incubator-mxnet#4856]
Consider the DSD paper:
https://www.arxiv.org/abs/1607.04381

Basically the steps are:
1) Train.
2) Find weights close to 0. Set them to 0, and set their learning rate to 0. 
3) Train. Unfreeze all weights. Train again.

**In theory, this is possible by using set_lr_mult in optimizer api, yet it might be really cumbersome and slow: there'll be a very long list of "freezed weights" and I suppose the optimizer will go through the whole list for each parameter?**

**Or does mxnet store the learning rate of each weight so there'll be no need to search? In that case, can we go through all weights and change the learning rate of each weight directly, instead of feeding a very long list to the optimizer?**

**Comment by piiswrong** [comment]
see mx.mod.Module(fixed_param_names)

**Comment by cheshirecats** [comment]
@piiswrong Thanks. The problem is, how to let fixed_param_names be those weight that are close to 0.

**Comment by Guneet-Dhillon** [comment]
@cheshirecats you can have a look at this https://github.com/apache/incubator-mxnet/tree/master/example/dsd

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!
Also, do please check out our [forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) for general "how-to" questions.
